cypress/integration/new_group_page.js

```diff
@@ -0,0 +1,107 @@
+describe('New Group page tests', () => {
+  context('no access before login', () => {
+    it('redirects to login', () => {
+      cy.visit('/groups/new');
+      cy.location('pathname').should('eq', '/login');
+    });
+  });
+
+  context('authenticated', () => {
+    beforeEach(() => {
+      cy.fixture('devices_detailed').as('devices');
+      cy.fixture('group.first-floor.created').as('postNewGroupResponse');
+      cy.server();
+      cy.route('GET', '/appengine/v1/*/devices?details=true', '@devices');
+      cy.route({
+        method: 'POST',
+        url: '/appengine/v1/*/groups',
+        status: 201,
+        response: '@postNewGroupResponse',
+      }).as('postNewGroup');
+      cy.login();
+      cy.visit('/groups/new');
+    });
+
+    it('successfully loads New Group page', () => {
+      cy.location('pathname').should('eq', '/groups/new');
+      cy.get('h2').contains('Create a New Group');
+    });
+
+    it('displays devices list correctly', function () {
+      cy.get('.main-content').within(() => {
+        cy.get('table tbody').find('tr').should('have.length', this.devices.data.length);
+        this.devices.data.forEach((device, index) => {
+          cy.get(`table tbody tr:nth-child(${index + 1})`).within(() => {
+            cy.contains(device.id);
+            Object.values(device.aliases).forEach((alias) => {
+              cy.contains(alias);
```

Comment:
not sure this cy context points to the elemet expressed in the forEach parameter. I think it's better to use the `expect( var )` function

Reply:
Maybe I didn't get your concern but this is valid Cypress code. Also, `cy.something()` functions are more idiomatic than `expect()`; the latter usually complicates tests as it requires direct DOM manipulation, which is something you'd like to avoid in a functional test

Reply:
This `cy.contains` is called inside a callback which itself is wrapped in a `within` block on a list of `tr` elements.
Are you sure it references the row from the actual within context?

When used directly from the cy object, it searches the whole document
https://docs.cypress.io/api/commands/contains.html#Usage

Reply:
The `cy.contains()` won't search the whole document since it is scoped in a `cy.within()`, which scopes all subsequent cy commands to within the element it refers to.
https://docs.cypress.io/api/commands/within.html#Tables

`within()` fixes the scope to `tr:nth-child(${index + 1})`, which is a single specific row.